opensearch-fetcher/log_fetching_service.py

```diff
@@ -0,0 +1,137 @@
+
+ES_ENDPOINT = os.environ["ES_ENDPOINT"]
+ES_USERNAME = os.environ["ES_USERNAME"]
+ES_PASSWORD = os.environ["ES_PASSWORD"]
+TIME_RANGE_SECONDS = int(os.getenv("TIME_RANGE_SECONDS", "10"))
+
+nw = NatsWrapper()
+
+async def send_all_results_to_nats(es, start_ts, end_ts):
+    scroll_value = "1m"
+    es_query_body = {
+        "query": {
+            "bool": {
+                "filter": [{"range": {"time": {"gte": start_ts, "lt": end_ts}}}]
+            },
+        }
+    }
+
+    first_page = True
+    while True:
+        if first_page:
+            current_page = await es.search(index="logs", body=es_query_body, scroll=scroll_value, size=2000)
+            first_page = False
+        else:
+            current_page = await es.scroll(scroll_id=scroll_id, scroll=scroll_value)
+        if "_scroll_id" in current_page:
+            scroll_id = current_page["_scroll_id"]
+            current_page_results = current_page["hits"]["hits"]
+            if len(current_page_results) > 0:
+                accumulated_results = []
+                for result in current_page_results:
+                    result_dict = result["_source"].copy()
+                    result_dict["_id"] = result["_id"]
+                    accumulated_results.append(result_dict)
+                accumulated_results_df = pd.DataFrame(accumulated_results)
+                logging.info("Published {} logs to Nats now".format(len(accumulated_results)))
+                await nw.publish("raw_logs", accumulated_results_df.to_json().encode())
+            else:
+                break
+        else:
+            break
+
+async def fetch_logs():
+    es = await setup_es_connection()
+    query_time = 0
+    try:
+        current_ts = int((datetime.now().timestamp() - TIME_RANGE_SECONDS) * 1000)
+        last_fetched_exists = await es.indices.exists("last_fetched")
+        last_fetched_timestamp_results = None
+        # On startup, check if index last_fetched already exists. Create it if it doesn't.
+        if last_fetched_exists:
+            last_fetched_timestamp_results = (await es.search(index="last_fetched", body={"query": {"match_all": {}}}))["hits"]["hits"][0]
+            last_fetched_timestamp = last_fetched_timestamp_results["_source"]["last_fetched_timestamp"]
+            await send_all_results_to_nats(es, last_fetched_timestamp, current_ts)
+        else:
+            await es.index(index="last_fetched", body={"last_fetched_timestamp": current_ts})
+
+        while True:
+            await asyncio.sleep(TIME_RANGE_SECONDS - min(query_time, TIME_RANGE_SECONDS))
+            end_ts = current_ts + (TIME_RANGE_SECONDS * 1000)
+            start_time = time.time()
+            await send_all_results_to_nats(es, current_ts, end_ts)
+            if last_fetched_timestamp_results:
+                await es.update(index="last_fetched", doc_type=last_fetched_timestamp_results["_type"], id=last_fetched_timestamp_results["_id"], body={"doc": {"last_fetched_timestamp": current_ts}})
```

Comment:
1.  it's better to use `index()` than `update()`. in OpenSearch, operation `update()` == `index()` + `delete()` so it's much more expensive.
       
2.`end_ts` should be the timestamp inserted to OpenSearch , as it will be the start_ts for line 67

Reply:
Index will create several new rows over time. We just need one row in this "last_fetched" index.